xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -131,62 +136,160 @@ XdsClient createXdsClient() {
-            + "    }\n"
-            + "  ]\n"
-            + "}";
-        Map<String, ?> config;
-        try {
-          config = (Map<String, ?>) JsonParser.parse(serviceConfig);
-        } catch (IOException e) {
-          listener.onError(
-              Status.UNKNOWN.withDescription("Invalid service config").withCause(e));
-          return;
+            "Received config update with {0} routes from xDS client {1}",
+            update.getRoutes().size(),
+            xdsClient);
+        rawLbConfig = generateXdsRoutingRawConfig(update.getRoutes());
+      } else {
+        Route defaultRoute = Iterables.getOnlyElement(update.getRoutes());
+        String clusterName = defaultRoute.getRouteAction().getCluster();
+        if (!clusterName.isEmpty()) {
+          logger.log(
+              XdsLogLevel.INFO,
+              "Received config update from xDS client {0}: cluster_name={1}",
+              xdsClient,
+              clusterName);
+          rawLbConfig = generateCdsRawConfig(clusterName);
+        } else {
+          logger.log(
+              XdsLogLevel.INFO,
+              "Received config update with one weighted cluster route from xDS client {0}",
+              xdsClient);
+          List<ClusterWeight> clusterWeights = defaultRoute.getRouteAction().getWeightedCluster();
+          rawLbConfig = generateWeightedTargetRawConfig(clusterWeights);
         }
-        logger.log(XdsLogLevel.INFO, "Generated service config:\n{0}", serviceConfig);
-        Attributes attrs =
-            Attributes.newBuilder()
-                .set(XdsAttributes.XDS_CLIENT_POOL, xdsClientPool)
-                .build();
-        ConfigOrError parsedServiceConfig = serviceConfigParser.parseServiceConfig(config);
-        ResolutionResult result =
-            ResolutionResult.newBuilder()
-                .setAddresses(ImmutableList.<EquivalentAddressGroup>of())
-                .setAttributes(attrs)
-                .setServiceConfig(parsedServiceConfig)
-                .build();
-        listener.onResult(result);
       }
 
-      @Override
-      public void onError(Status error) {
-        // In order to distinguish between IO error and resource not found, which trigger
-        // different handling, return an empty resolution result to channel for resource not
-        // found.
-        // TODO(chengyuanzhang): Returning an empty resolution result based on status code is
-        //  a temporary solution. More design discussion needs to be done.
-        if (error.getCode().equals(Code.NOT_FOUND)) {
-          logger.log(
-              XdsLogLevel.WARNING,
-              "Received error from xDS client {0}: {1}", xdsClient, error.getDescription());
-          listener.onResult(ResolutionResult.newBuilder().build());
-          return;
+      Map<String, ?> serviceConfig =
+          ImmutableMap.of("loadBalancingConfig", ImmutableList.of(rawLbConfig));
+      logger.log(XdsLogLevel.INFO, "Generated service config:\n{0}", serviceConfig);
```

Comment:
This message would be almost unreadable, as `serviceConfig` is a deeply nested map. We may need a utility that coverts `Map<String, ?>` to `String`, but seems would only be useful for visualization. It's nontrivial to implement that. 

So maybe, we have to delete this log message (or change it to `DEBUG`?) to avoid printing out tons of hard-to-read data.

Reply:
Service config is in general a very large, as it contains not only lb config.  I think service config is important logging information and it might worth INFO level. But we need to de-dup ConfigUpdate to avoid logging spam.

Reply:
I mean, printing out the `Map<String, ?>` is bad, it's unreadable when it's large. If we have utility to convert `Map<String, ?>` to a nicely formatted JSON string, that would be the best. But, we don't have such a utility.

Reply:
Use gson reverse parser now.